### ISSUE TITLE
Find alternative to xclip for wayland

### ISSUE THREAD

**Issue by keis** [uzbl/uzbl#337]
`xclip` is used extensively in the example config but this does not always work on wayland. some compositors like gnome will sync the clipboard between X but this is not something we can rely on.

To my knowledge there is no cross desktop spec for clipboards yet so this would either have to 
a) wait for something like that to emerge
or
b) we could implement commands/variables to access to clipboard in uzbl-core

**Comment by mathstuf** [comment]
Fedora says that the protocol for this [is pending](https://fedoraproject.org/wiki/Wayland_features#Clipboard_manager) (last edited September 1 2016). There is [a protocol section](https://wayland.freedesktop.org/docs/html/ch04.html#sect-Protocol-data-sharing) for a clipboard though.

**Comment by keithbowes** [comment]
Perhaps you could use the [JavaScript clipboard API](https://developer.apple.com/library/content/documentation/AppleApplications/Conceptual/SafariJSProgTopics/CopyAndPaste.html) instead of an external util and just let Webkit take care of it regardless of the graphical environment.

**Comment by mathstuf** [comment]
No, because Uzbl executes JavaScript with DOM access at the same level as the page that is loaded. If you deny access to the page, you deny access to Uzbl as well.

**Comment by keithbowes** [comment]
That's unfortunate.  I thought it might be possible to use it in a clean context.  It is true that WebKit seems not to be up to the [W3C spec](https://www.w3.org/TR/clipboard-apis/#clipboard-event-interfaces).  The following code produces the two expected message boxes in Firefox:
```
try
{
  document.addEventListener('copy', function(e) { alert('The text "' + e.clipboardData.getData('text/plain') + '" was copied.');});
  document.addEventListener('paste', function(e) { alert('The text "' + e.clipboardData.getData('text/plain') + '" was pasted.');});
  ev = new ClipboardEvent('copy');
  ev.clipboardData.setData('text/plain', 'Hello World');
  document.dispatchEvent(ev);
  ev.type = 'paste';
  document.dispatchEvent(ev);
}
catch (e)
{
  document.write('Exception caught: ' + e.name + ': ' + e.message + '<br/>');
}
```

But in Uzbl, I get:
>> Exception caught: ReferenceError: Can't find variable: ClipboardEvent

Of course, I thought that might just be because of my distro's old version of WebKit (2.4.10).

**Comment by bugaevc** [comment]
Hi, I wrote [wl-clipboard](https://github.com/bugaevc/wl-clipboard) as a Wayland alternative to xclip, maybe you'll find it useful.

**Comment by keis** [comment]
@bugaevc cool stuff, is this using some official wayland clipboard protocol or is it more a gtk tool which IIRC was the line pursued by the gnome folks before?

**Comment by bugaevc** [comment]
It absolutely uses the official Wayland protocol for the core functionality and should work with any Wayland compositor.

_In addition to that_, it can make use of the GTK+ Wayland protocol extension for primary clipboard, _if_ the compositor supports it (e.g. Mutter & wlroots-based compositors do). To clarify, this is just a Wayland protocol extension that GTK+ (as well as others) implement that originated in GTK+ and has `gtk_` in its name; wl-clipboard does not link to GTK+ or anything.
